L56-objc-bazel-support.md

```diff
@@ -0,0 +1,202 @@
+* All the necessary external dependencies are loaded with `grpc_deps()` in `//bazel:grpc_deps.bzl` and are hidden from users.
+
+### Rules for Compiling `.proto` Files
+
+The gist of these custom rules is to run protobuf compiler and Objective-C plugin on provided `.proto` files with `ctx.action.run`. Those executables are available from `@com_google_protobuf//:protoc` and `@com_github_grpc_grpc//:grpc_objective_c_plugin`.
+
+We use the native `proto_library` rule as a manager for `.proto` files (i.e. their package paths and dependencies). They wrap the `.proto` files and are passed into `objc_proto_grpc_library` and `objc_grpc_library` as `deps`.
+
+`objc_grpc_library`, takes in as `deps` a list of `proto_library` targets and geneates the message stubs (excluding the service ones) for these targets and all their transitively dependent protos. In addition, it takes in a list of labels of `.proto` files as `srcs`. If the `.proto` file is in the same package, users can use its relative path. The list of `.proto` files should all contain service stubs; otherwise Bazel will complain about certain `.pbrpc.{h,m}` files not being generated. `objc_grpc_library`, generates services stubs for a `.proto` file if and only if the `.proto` file is listed in `srcs`.
+
+As a result of compiling every `.proto` files in the dependency chain, the app-specific code only needs to depend on the one or the few `objc_grpc_library`'s at the bottom of the dependency graph.
+
+For `objc_grpc_library`, it is also possible to tell that [well known protos](https://github.com/protocolbuffers/protobuf/tree/master/src/google/protobuf) are required in the dependency, by passing `True` for the field `use_well_known_protos`.
+
+In terms of the execution of `protoc`, the command is similar to that in a [podspec](https://github.com/grpc/grpc/blob/0803c79411597f58eae0b12b4eb272c506b8cdbb/examples/objective-c/helloworld/HelloWorld.podspec) (all `.proto` targets in `deps`, including their transitive dependencies, are provided as inputs, and their directory from the `WORKSPACE` root are added to `-I` flags programmatically). The output directory is set to `//bazel-out/<*CPU architecture*>/bin/<package name>/_generated_protos/` so that bazel is able to locate the generated files.
+
+Generated files in the directory above follow the same hierarchy as the `.proto` files. Consider this project where we are building from package `//A`:
+
+![hierarchy Example](L56_graphics/hierarchy1.png)
+
+The resulting structure in `bin` will be:
+
+![hierarchy Result Example](L56_graphics/hierarchy2.png)
+
+Lastly, three other targets are created to split the files into `hdrs`, `srcs` (potentially empty, since there might be no service stubs), and `non_arc_srcs`. They will be fed into a `objc_library` rule.
+
+When `#import`-ing `.proto` and `.pb*.{h,m}` files, always use their *absolute paths* from the `WORKSPACE` root.
+
+
+### Compatibility with Google3
```

Comment:
Ditto. I would suggest removing this part on Github proposal, and open an internal gDoc counterpart of this proposal that include this section

Reply:
Removed the part discussing implementation in Google3.